risc0/zkvm/src/prove/dev_mode.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::collections::BTreeMap;
+
+use anyhow::Result;
+use risc0_binfmt::MemoryImage;
+
+use crate::{
+    prove::Prover,
+    receipt::{InnerReceipt, Receipt, VerifierContext},
+    Executor, ExecutorEnv, Segment, SegmentReceipt, Session,
+};
+
+/// An implementation of a [Prover] for development and testing purposes.
```

Comment:
Can we expand this documentation a bit? Since running intended-to-be-secure code in DevMode would be such a substantial failure, I want developers to have every opportunity to learn when to use it (and when not)
```suggestion
/// An implementation of a [Prover] for development and testing purposes.
///
/// When code is "proven" with this DevModeProver, no actual proof is produced.
/// Instead, the guest code is executed and a fake receipt with the true journal
/// but no cryptographic information is returned. This receipt will verify with
/// if DevMode is turned on, but will otherwise fail to verify.
///
/// This allows for rapid development by quickly iterating on code without
/// forcing the developer to wait for proving to complete. However, it must not
/// be used in production as it provides no security whatsoever.
///
/// DevMode is only used when the environment variable `DEV_MODE` is set. It can
/// also be fully disabled at compile time, regardless of environment variables,
/// by setting the feature flag `disable-dev-mode` on the `risc0_zkvm` crate.
```

Reply:
```suggestion
/// An implementation of a [Prover] for development and testing purposes.
/// This DevModeProver does not produce an actual proof.
/// Instead, the guest code is executed and a fake receipt is returned with
/// accurate journal contents but no cryptographic information.
/// Because the receipt is fake, a verifier can only "verify" this receipt
/// if DevMode is turned on; verification will otherwise fail.
///
/// CONVENIENT, BUT NOT MEANT FOR PRODUCTION
/// Dev mode supports rapid development by allowing the developer to quickly iterate
/// on code without being forced to wait for proving to complete. 
/// However, it must not be used in production as it provides no security whatsoever.
///
/// How to enable and disable dev mode:
/// Dev mode is only used when the environment variable `DEV_MODE` is set. It can
/// be fully disabled at compile time, regardless of environment variables,
/// by setting the feature flag `disable-dev-mode` on the `risc0_zkvm` crate.
```

Reply:
I'm happy with your rewrite, but please add a line break between the first (summary) line and the in-depth documentation.

(If this isn't just a typo & you're not familiar, Rust will use everything before the first empty line as a summary, so it's important to put a full empty line in early. Best practices are to keep the summary single line. See https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components for more information.)